Modules/posixmodule.c

```diff
@@ -5490,6 +5490,11 @@ os__path_abspath_impl(PyObject *module, PyObject *path)
         result = posix_getcwd(0);
         goto exit;
     }
+    if (wcslen(abs) != abs_len) {
+        PyErr_Format(PyExc_ValueError,
+                     "_path_abspath: embedded null character in path");
+        goto exit;
```

Comment:
I'm not sure how I feel about only using the fallback for some but not all paths that have embedded null characters, depending on whether the problematic components get removed by `normpath()`.

Reply:
It's faster for paths with many `..`'s, and I don't think there's a good reason to raise an error if we can handle it correctly.

Reply:
The fallback implementation doesn't support per-drive working directories and fails to return an absolute path in drive-relative cases. It also doesn't reserve DOS device names in the final component (only the "nul" device in Windows 11, except for an unqualified device name without any path separators, such as "con"), or remove trailing spaces and dots from the final component. This means the result can vary significantly depending on whether a troublesome component with an embedded null character gets removed when the path is normalized.

Offhand I can't remember why `normpath()` gets called first. The fact that it removes an initial "." component is even more of a nuisance now. It prevents qualified referencing of files with certain DOS device names in the working directory such as "./con" versus unqualified "con". Support for the latter was added in Windows 11. For example:

```pycon
>>> nt._getfullpathname('./con')
'C:\\Temp\\con'
>>> nt._getfullpathname('con')
'\\\\.\\con'
>>> os.path.abspath('./con')
'\\\\.\\con'
```

Calling `normpath()` first was always an issue for unambiguous references to file streams of single-letter filenames in the current directory. For example:

```pycon
>>> nt._getfullpathname('./Z:spam')
'C:\\Temp\\Z:spam'
>>> os.path.abspath('./Z:spam')
'Z:\\spam'
```

Windows would beg to differ about the significance of a leading "." component, in these cases and others. Anyway, I thought this version could at least address the inconsistent behavior with embedded null characters, while still calling `normpath()` first.

Reply:
Maybe a parameter could be added to `_Py_normpath_and_size()` to make it skip past a leading "." component, but keep it in the result. This would retain support for the special-cased behavior that's implemented by `GetFullPathNameW()`.

Reply:
> Offhand I can't remember why normpath() gets called first.

```none
======================================================================
ERROR: test_abspath (__main__.TestNtpath.test_abspath)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\wanne\cpython\speedup-posixpath.abspath\Lib\test\test_ntpath.py", line 813, in test_abspath
    tester('ntpath.abspath("//../../../..")',  "\\\\..\\..\\")
    ~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\wanne\cpython\speedup-posixpath.abspath\Lib\test\test_ntpath.py", line 60, in tester
    raise TestFailed("%s should return: %s but returned: %s" \
          %(str(fn), str(wantResult), str(gotResult)))
test.support.TestFailed: ntpath.abspath("//../../../..") should return: \\..\..\ but returned: \\..\
```

Is `GetFullPathNameW()` wrong? That seems strange.

Reply:
I recall now that there was an in-depth discussion about handling UNC edge cases. What do you think about the extra parameter for `_Py_normpath_and_size()` that I proposed? It would allow normalizing the path first while keeping the system behavior for a leading "." component.

> Is `GetFullPathNameW()` wrong? That seems strange.

IMO, what `GetFullPathNameW()` does is superficially wrong, but it's okay in practice since ".." can't be a valid share name, and it doesn't have to retain an explicit root on a UNC share since the root is implicit.

What's happening here is that trailing dots are removed at the end by code that has nothing to do with handling ".." components. For example, notice how the API function removes the explicit root in the following case: 

```pycon
>>> nt._getfullpathname('//x/y/..')
'\\\\x\\y'
```
Then a share component that's nothing but dots and spaces will be subject to normal trimming. For example:

```pycon
>>> nt._getfullpathname('//x/./..')
'\\\\x\\'
>>> nt._getfullpathname('//x/../..')
'\\\\x\\'
>>> nt._getfullpathname('//x/.../..')
'\\\\x\\'
>>> nt._getfullpathname('//x/. . ./..')
'\\\\x\\'
```

But it's retained if it's not the final component:

```pycon
>>> nt._getfullpathname('//x/./../z')
'\\\\x\\.\\z'
>>> nt._getfullpathname('//x/../../z')
'\\\\x\\..\\z'
>>> nt._getfullpathname('//x/.../../z')
'\\\\x\\...\\z'
>>> nt._getfullpathname('//x/. . ./../z')
'\\\\x\\. . .\\z'
```

Reply:
> What do you think about the extra parameter for `_Py_normpath_and_size()` that I proposed? It would allow normalizing the path first while keeping the system behavior for a leading "." component.

That's implemented, but could we modify the main loop to use an explicit curdir for cases with '..' too?

```python
>>> import nt
>>> nt._path_abspath('con')
'\\\\.\\con'
>>> nt._path_abspath('./con') 
'C:\\Users\\wanne\\cpython\\speedup-posixpath.abspath\\con'
>>> nt._path_abspath('foo/../con')
'\\\\.\\con'
```

Ideally like this:

```none
'foo' / '..'    -> '.'
'.' / '..'      -> '..'
'.' / 'foo'     -> './foo'
'./foo' / '..'  -> '.'
'./foo' / 'bar' -> 'foo/bar'
```

Reply:
Probably too much trouble, though. And the use case for regular users is very minor.

Reply:
I think it's worthwhile for "foo/../con" to normalize as "./con" when returning an explicit current directory for internal use. It ensures that the `GetFullPathNameW()` result will match what it would have been if it were called directly on the path. This would be implemented generally for all relative paths that initially have at least one path separator. Manually start the normalized result with "./" or ".\\".

Note that the behavior for normalizing drive-relative paths is currently wrong in this case on Windows:

```pycon
>>> nt._path_normpath('C:./spam')
'C:.\\spam'
```
That's the result that we want here when returning an explicit current directory for *internal use*, but that's not the expected behavior for `normpath()`. The fallback implementation that's used on POSIX gets it right:

```pycon
>>> ntpath.normpath('C:./spam')
'C:spam'
```

Reply:
> Note that the behavior for normalizing drive-relative paths is currently wrong in this case on Windows:

Fixed that inconsistency.

> I think it's worthwhile for "foo/../con" to normalize as "./con" when returning an explicit current directory for internal use.

I initially thought about doing it in the loop, but that looks too complicated. It should be fairly easy to do it afterwards.

```none
'foo/..' -> ''
 |   +- p2
 +----- p3

'./..' -> '..'
 | +- p2
 +--- p3

'./foo' (unchanged)

'./foo/..' -> './'
   |   +- p2
   +----- p3

'./foo/bar' -> 'foo/bar'
 |     +- p2
 +------- p3 (needs to be computed)

'foo/bar/..' -> './foo/'
 |   |   +- p2
 |   +----- p3
 +--------- p4 (needs to be computed)
```

Reply:
"./foo/.." should normalize as ".", not "./". With an explicit current directory, "foo/bar/.." should normalize as "./foo", not "./foo/".

A "." component only needs to be prefixed if the input path has a path separator and the result doesn't have a path separator. Also, that's only if the buffer size allows it. "foo/" has to normalize as "foo" because there isn't space for "./foo".

Reply:
> "./foo/.." should normalize as ".", not "./".

Yeah, after the loop that trailing slash will be removed.

> A "." component only needs to be prefixed if the input path has a path separator and the result doesn't have a path separator.

I'm counting the number of explicit parts now, so 'foo/' isn't qualified, but `foo/.` is.

Reply:
This could be an option that's worth exposing in `normpath()`. In a search context on POSIX and Windows, "spam" differs from "./spam". On Windows, "spam/eggs" differs from "./spam/eggs" in a search context, and "C:spam" differs from "./C:spam" in an open context.

Reply:
But don't jump on that discussion yet on the ideas list. Get it accepted as an internal option that's already implemented before trying to extend `normpath()`. It's more work than just exposing the parameter because the fallback implementation also has to change.

Reply:
Does it look correct now?

```none
'./foo/bar' (unchanged)
'foo/../bar' -> './bar'
'foo/bar/..' -> './foo'
'foo/.' -> './foo'
```